File: apps/api/src/application/import/junit.parser.ts
import { TestResultStatus } from "@testcraft/types";
import { XMLParser } from "fast-xml-parser";

import { DomainError } from "@/domain/errors";

import type { ParsedStep, ParsedTestCase } from "./import.repository";

const ARRAY_NODES = new Set(["testsuite", "testcase"]);

const junitParser = new XMLParser({
  ignoreAttributes: false,
  attributeNamePrefix: "",
  textNodeName: "#text",
  isArray: (name) => ARRAY_NODES.has(name),
});

const extractXmlText = (node: unknown): string | null => {
  if (typeof node === "string") return node || null;
  if (typeof node === "object" && node !== null) {
    const obj = node as Record<string, unknown>;
    const text = obj.message ?? null;

    return typeof text === "string" ? text.trim() || null : null;
  }

  return null;
};

const resolveStatus = (
  testcase: Record<string, unknown>,
): { status: TestResultStatus; notes: string | null } => {
  if ("failure" in testcase || "error" in testcase)
    return {
      status: TestResultStatus.Failed,
      notes: extractXmlText(testcase.failure ?? testcase.error),
    };
  if ("skipped" in testcase)
    return { status: TestResultStatus.Skipped, notes: null };

  return { status: TestResultStatus.Passed, notes: null };
};

const strVal = (value: unknown): string | null =>
  typeof value === "string" && value ? value : null;

const resolveSuiteName = (
  suiteNameFromAttr: string | null,
  testcase: Record<string, unknown>,
): string => {
  const classname = strVal(testcase.classname);

  if (suiteNameFromAttr?.toLowerCase().endsWith(".dll")) {
    return classname ?? suiteNameFromAttr;
  }

  return suiteNameFromAttr ?? classname ?? "Default Suite";
};

const SEP = " > ";

const parseSteps = (caseName: string): ParsedStep[] => {
  const idx = caseName.lastIndexOf(SEP);

  if (idx === -1) return [];

  return [
    {
      order: 1,
      action: caseName.slice(0, idx),
      expectedResult: caseName.slice(idx + SEP.length),
    },
  ];
};

const DEFAULT_RUN_NAME = "Imported Run";

export const parseJUnit = (
  xml: string,
): { runName: string; cases: ParsedTestCase[] } => {
  let doc: Record<string, unknown>;

  try {
    doc = junitParser.parse(xml) as Record<string, unknown>;
  } catch {
    throw new DomainError("Invalid JUnit XML: could not parse the document");
  }

  let suites: Record<string, unknown>[] = [];
  let runName = DEFAULT_RUN_NAME;

  if (doc.testsuites) {
    const testsuites = doc.testsuites as Record<string, unknown>;

    runName = strVal(testsuites.name) ?? DEFAULT_RUN_NAME;
    suites = (testsuites.testsuite as Record<string, unknown>[]) ?? [];
  } else if (doc.testsuite) {
    suites = doc.testsuite as Record<string, unknown>[];
    const first = suites[0];

    if (first) runName = strVal(first.name) ?? DEFAULT_RUN_NAME;
  }

  const cases = suites.flatMap((suite) => {
    const suiteNameFromAttr = strVal(suite.name);

    return ((suite.testcase as Record<string, unknown>[]) ?? []).map(
      (testcase) => {
        const caseName = strVal(testcase.name) ?? "Unknown";

        return {
          suiteName: resolveSuiteName(suiteNameFromAttr, testcase),
          caseName,
          ...resolveStatus(testcase),
          steps: parseSteps(caseName),
        };
      },
    );
  });

  return { runName, cases };
};

File: apps/Api/src/TestCraft.Application/Import/JUnitParser.cs
using System.Xml;
using System.Xml.Linq;
using TestCraft.Application.Common.Exceptions;
using TestCraft.Domain.Enums;

namespace TestCraft.Application.Import;

public static class JUnitParser
{
    private const string DefaultRunName = "Imported Run";
    private const string Separator = " > ";

    public static (string RunName, List<ParsedTestCase> Cases) Parse(string xml)
    {
        XDocument doc;
        try
        {
            var settings = new XmlReaderSettings { DtdProcessing = DtdProcessing.Prohibit };
            using var reader = XmlReader.Create(new StringReader(xml), settings);
            doc = XDocument.Load(reader);
        }
        catch (Exception)
        {
            throw new DomainException("Invalid JUnit XML: could not parse the document");
        }

        var root = doc.Root;
        if (root is null)
        {
            throw new DomainException("Invalid JUnit XML: could not parse the document");
        }

        string runName;
        IEnumerable<XElement> suites;

        if (root.Name.LocalName == "testsuites")
        {
            runName = StrVal(root.Attribute("name")?.Value) ?? DefaultRunName;
            suites = root.Elements("testsuite");
        }
        else if (root.Name.LocalName == "testsuite")
        {
            suites = [root];
            runName = StrVal(root.Attribute("name")?.Value) ?? DefaultRunName;
        }
        else
        {
            runName = DefaultRunName;
            suites = [];
        }

        var cases = new List<ParsedTestCase>();
        foreach (var suite in suites)
        {
            var suiteNameFromAttr = StrVal(suite.Attribute("name")?.Value);
            foreach (var testcase in suite.Elements("testcase"))
            {
                var caseName = StrVal(testcase.Attribute("name")?.Value) ?? "Unknown";
                var suiteName = ResolveSuiteName(suiteNameFromAttr, testcase);
                var (status, notes) = ResolveStatus(testcase);

                var durationMs = ParseDurationMs(testcase.Attribute("time")?.Value);

                cases.Add(
                    new ParsedTestCase
                    {
                        SuiteName = suiteName,
                        CaseName = caseName,
                        Status = status,
                        Notes = notes,
                        DurationMs = durationMs,
                        Steps = ParseSteps(caseName),
                    }
                );
            }
        }

        return (runName, cases);
    }

    private static (TestResultStatus Status, string? Notes) ResolveStatus(XElement testcase)
    {
        var failure = testcase.Element("failure") ?? testcase.Element("error");
        if (failure is not null)
        {
            return (TestResultStatus.Failed, ExtractXmlText(failure));
        }

        if (testcase.Element("skipped") is not null)
        {
            return (TestResultStatus.Skipped, null);
        }

        return (TestResultStatus.Passed, null);
    }

    private static string? ExtractXmlText(XElement element)
    {
        var messageAttr = element.Attribute("message");
        if (messageAttr is not null)
        {
            var text = messageAttr.Value.Trim();
            if (text.Length > 0)
            {
                return text;
            }
        }

        var raw = element.Value.Trim();

        return string.IsNullOrEmpty(raw) ? null : raw;
    }

    private static List<ParsedStep> ParseSteps(string caseName)
    {
        var separatorIndex = caseName.LastIndexOf(Separator, StringComparison.Ordinal);
        if (separatorIndex == -1)
        {
            return [];
        }

        return
        [
            new ParsedStep
            {
                Order = 1,
                Action = caseName[..separatorIndex],
                ExpectedResult = caseName[(separatorIndex + Separator.Length)..],
            },
        ];
    }

    private static long? ParseDurationMs(string? timeStr)
    {
        if (string.IsNullOrEmpty(timeStr))
            return null;
        if (
            double.TryParse(
                timeStr,
                System.Globalization.NumberStyles.Any,
                System.Globalization.CultureInfo.InvariantCulture,
                out var seconds
            )
        )
            return (long)(seconds * 1000);
        return null;
    }

    private static string? StrVal(string? value) => string.IsNullOrEmpty(value) ? null : value;

    private static string ResolveSuiteName(string? suiteNameFromAttr, XElement testcase)
    {
        var classname = StrVal(testcase.Attribute("classname")?.Value);

        if (
            suiteNameFromAttr is not null
            && suiteNameFromAttr.EndsWith(".dll", StringComparison.OrdinalIgnoreCase)
        )
        {
            return classname ?? suiteNameFromAttr;
        }

        return suiteNameFromAttr ?? classname ?? "Default Suite";
    }
}
